common/lib/core-interfaces/test-d/index.test-d.ts

```diff
@@ -0,0 +1,88 @@
+ * Licensed under the MIT License.
+ */
+/* eslint-disable unicorn/filename-case */
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {expectError} from "tsd";
+import { IFluidLoadable, IProvideFluidLoadable, FluidObject, FluidObjectKeys } from "../dist";
+
+
+declare function getUnknownFluidObject(): FluidObject;
+
+declare function useUnknownFluidObject(params: FluidObject | undefined): void;
+
+declare function useProvider<T extends FluidObject>(params: FluidObject<T> | undefined): void;
+
+declare function useProviderKey<T,TKey extends FluidObjectKeys<T> = FluidObjectKeys<T>>(key: TKey): void;
+
+declare function useLoadable(params: FluidObject<IFluidLoadable> | undefined): void;
+
+// test implicit conversions between FluidObject and a FluidObject with a provides interface
+{
+    const FluidObject: FluidObject<IProvideFluidLoadable> = getUnknownFluidObject();
+    useUnknownFluidObject(FluidObject);
+    useUnknownFluidObject(FluidObject.IFluidLoadable);
+    useProvider(FluidObject);
+    useProvider(FluidObject.IFluidLoadable);
+    useLoadable(FluidObject);
+    useLoadable(FluidObject.IFluidLoadable);
+    expectError(FluidObject.handle);
+    FluidObject.IFluidLoadable?.handle;
+    const unknown: FluidObject | undefined = FluidObject.IFluidLoadable;
+    useUnknownFluidObject(unknown);
+    useProvider(unknown);
+    useProvider<IFluidLoadable>(unknown);
+    useLoadable(unknown);
+}
+
+// test implicit conversions between FluidObject and a FluidObject with a implementation interface
+{
+    const foo: FluidObject<IFluidLoadable> = getUnknownFluidObject();
+    useUnknownFluidObject(foo);
+    useUnknownFluidObject(foo.IFluidLoadable);
+    useProvider(foo);
+    useProvider(foo.IFluidLoadable);
+    useLoadable(foo);
+    useLoadable(foo.IFluidLoadable);
+    expectError(foo.handle);
+    foo.IFluidLoadable?.handle;
+    const unknown: FluidObject | undefined = foo.IFluidLoadable;
+    useUnknownFluidObject(unknown);
+    useProvider(unknown);
+    useProvider<IFluidLoadable>(unknown);
+    useLoadable(unknown);
+}
+
+// test getting keys
+{
+    useProviderKey<IProvideFluidLoadable>(IFluidLoadable);
+    useProviderKey<IFluidLoadable>(IFluidLoadable);
+    const loadableKey: keyof IFluidLoadable = "handle";
+    expectError(useProviderKey<IFluidLoadable>(loadableKey));
+}
+
+// test implicit conversions between FluidObject and a FluidObject with a partial provider interface
+{
+
+    interface IProvideFoo{
+        IFoo: IFoo;
+    }
+    interface IFoo extends Partial<IProvideFoo>{
+        doFoo();
+    }
+
+    const foo: FluidObject<IFoo> = getUnknownFluidObject();
+    useUnknownFluidObject(foo);
+    useUnknownFluidObject(foo.IFoo);
+    useProvider(foo);
+    useProvider(foo.IFoo);
+    foo.IFoo?.doFoo();
+    const fooKey: keyof IFoo = "doFoo";
+    expectError(useProviderKey<IFoo>(fooKey));
+    const unknown: FluidObject | undefined = foo.IFoo;
```

Comment:
Still don't get who would ever treat `foo.IFoo` as `FluidObject`. The following four lines don't care what's actually set here, can you change RHS to `getUnknownFluidObject()` or something?

Reply:
https://github.com/microsoft/FluidFramework/blob/2ea977e6d99449ac3e1db9da2d1bbb51f047186c/packages/loader/container-loader/src/loader.ts#L187

Reply:
that will eventually change to FluidObject

Reply:
there are also cases where a function may take a FluidObject, and internally inspect for one or more interfaces, and do something. I don't want the caller, who may have a provider or a strongly types object to have to cast to pass their object to such a function. this just validates that the typing supports and implicit cast.

Reply:
`The following four lines don't care what's actually set here` is exactly what i'm validating. this should be the case

Reply:
ok